protocol/auditlog.go

```diff
@@ -100,6 +116,7 @@ func (l ConiksAuditLog) Insert(addr string, signKey sign.PublicKey,
 	// create the new directory history
 	h = newDirectoryHistory(addr, signKey, snaps[0])
 
+	// FIXME: remove this check --> caller calls Audit() before this function
```

Comment:
I think there are multiple ways to approach this. `Insert` shouldn't do anything but do an actual insertion into the audit log. I don't understand why `newDirectoryHistory` *must* be called before `Audit`, those are independent operations. The current logic is: "Did the audit pass? Yes, so let's insert the snapshots into our log/update our log". This totally fits with my previous statement, what it's saying is that insertion/updates to the log should happen *after* auditing checks pass. This *is* the logic that's currently implemented.

Now, I do agree that passing the response message to `Insert` would require the implementor to de-serialize twice, and that's not elegant. My intention was to keep the function signatures consistent with the ones in `directory.go`, where `Insert` also takes the `Response`. but I never felt like this really made sense in the auditor's case.

So, given my previous statement about how `Audit()` should update the snapshots if the checks pass, `Audit()` should really be calling `Insert()` and `Update`, not the other way around. In fact, at the application level, it would really make more sense to have the auditor simply call `Audit()` on a new server response and have the insertion/update be abstracted away from the application layer. Then, we can unexport `Insert` and `Update`, and pass the snapshots directly instead of the `Response`. What do you think?